### PR TITLE
Implement IBM Cloud deployment

### PR DESCRIPTION
**Pull request by petr-balogh**
Added support for new platform IBM Cloud
- OCP deployment
- OCS deployment

Fixes: #3494

Signed-off-by: Petr Balogh <pbalogh@redhat.com>

### CODE REVIEW

**Comment by petr-balogh** [comment]
Verification job which will cover ocp and ocs deployment at once:
https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/15511/console

OCS deployment passed + tier1 execution is running here:
https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/15508

**Review by dahorak — DISMISSED** [review]

File: ocs_ci/deployment/deployment.py [hunk 1]
```diff
@@ -397,6 +410,16 @@ def deploy_ocs_via_operator(self):
...
+        if (
+            config.ENV_DATA["platform"] == constants.IBMCLOUD_PLATFORM
+            and not live_deployment
+        ):
+            csv.wait_for_phase("Installing", timeout=720)
+            logger.info("Sleeping for 30 seconds before applying SA")
+            time.sleep(30)
+            link_all_sa_and_secret(constants.OCS_SECRET, self.namespace)
+            logger.info("Deleting all pods in openshift-storage namespace")
+            exec_cmd(f"oc delete pod --all -n {self.namespace}")
```
Comment (jimdevops19) [inline ocs_ci/deployment/deployment.py:422]: Couldn't we use a `delete_all_pods` function here?

File: ocs_ci/deployment/ocp.py [hunk 2]
```diff
@@ -28,9 +28,10 @@ def __init__(self):
         self.deployment_type = config.ENV_DATA["deployment_type"].lower()
         if not hasattr(self, "flexy_deployment"):
             self.flexy_deployment = False
-        if not self.flexy_deployment:
+        if not self.flexy_deployment and self.deployment_platform != "ibm_cloud":
```
Comment (jimdevops19) [inline ocs_ci/deployment/ocp.py:31]: Should be  `!= constants.IBMCLOUD_PLATFORM` ?

File: ocs_ci/deployment/factory.py [hunk 3]
```diff
@@ -30,6 +31,7 @@ def __init__(self):
             "baremetal_upi": BAREMETALUPI,
             "gcp_ipi": GCPIPI,
             "powervs_upi": IBMDeployment,
+            "ibm_cloud_managed": IBMCloud,
```
Comment (jimdevops19) [inline ocs_ci/deployment/factory.py:34]: I see that we have all platforms classes was just curious what is the general purpose of `factory.py` :)

File: ocs_ci/deployment/deployment.py (hunk 1 above)
Comment (petr-balogh) [reply ocs_ci/deployment/deployment.py:422]: I don't need to do anything else with this and it's not worth to even create ocp object here in order to call it. So one line command is easier option here.

File: ocs_ci/deployment/factory.py (hunk 3 above)
Comment (petr-balogh) [reply ocs_ci/deployment/factory.py:34]: It's one of the common design pattern of object oriented programming ;) you can try to google it ;) https://www.tutorialspoint.com/design_pattern/factory_pattern.htm this is one of the first result.

File: ocs_ci/deployment/ocp.py (hunk 2 above)
Comment (petr-balogh) [reply ocs_ci/deployment/ocp.py:31]: yep, fixed

**Review by dahorak — APPROVED** [review]

**Review by jimdevops19 — APPROVED** [review]

File: docs/supported_platforms.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+Created At    2020-10-19T15:05+0000
+API Key       OUR_KEY
+Locked        false
+```
+
+### Deployment types
+
+#### Managed
+
+This deployment type is managed - there is no openshift-installer used for
```
Comment (clacroix12) [inline docs/supported_platforms.md:102]: Isn't this just UPI? What is the difference between managed and UPI?

Comment (petr-balogh) [reply docs/supported_platforms.md:102]: @clacroix12 the IBM Cloud is not installed by openshift-installer so there is no UPI or IPI kind of deployment. It's completely managed by IBM Cloud. So if you would like to have new machines attached to cluster you interact through IBM Cloud API or CLI and then IBM Cloud will add everything to the cloud.

For more details see here:
https://cloud.ibm.com/docs/openshift?topic=openshift-roks-overview

Basically IBM has some big cluster, it has it's own master nodes and you will get just dedicated space under this big cluster with own worker nodes. You don't have access to the masters but only to the workers.

I think that we can have this managed deployment type used also for OpenShift Dedicated later on. https://www.openshift.com/products/dedicated/
